RavenDB-21021 Optimizing streaming operations in Corax

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21021 
https://issues.hibernatingrhinos.com/issue/RavenDB-21131
https://issues.hibernatingrhinos.com/issue/RavenDB-21092
https://issues.hibernatingrhinos.com/issue/RavenDB-20553
https://issues.hibernatingrhinos.com/issue/RavenDB-21137
https://issues.hibernatingrhinos.com/issue/RavenDB-21141

### Additional description

Expose Corax query plan in the query output 

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.

ayende committed:
RavenDB-21021 Include Corax DebugView (expose query structure) in Timings()

ayende committed:
RavenDB-21092 Corax streaming optimization - query with just order by

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -216,9 +215,23 @@ internal static IQueryMatch BuildQuery(Parameters builderParameters, out OrderMe
             }
-            else
+            // just one item, with known field types
+            else if (sortMetadata is [{ FieldType: MatchCompareFieldType.Floating or MatchCompareFieldType.Integer or MatchCompareFieldType.Sequence } sortBy])
```
This is the core of the change for [RavenDB-21092](https://issues.hibernatingrhinos.com/issue/RavenDB-21092) Corax streaming optimization - query with just order by

The idea is that if we have an `order by` without a `where`, we just set the scan on the right field.

That is slightly different from the usual `StreamingOptimization` and I'm not sure if this is the right place for this.

maciejaszyk replied:
Looks good

maciejaszyk commented on src/Corax/Constants.cs:
```diff
@@ -6,6 +6,9 @@ namespace Corax
 {
     public static class Constants
     {
+        public const string BeforeAllKeys = "BeforeAllKeys-a8e5f221-613e-4eae-9962-2689e7c44506";
+        public const string AfterAllKeys = "AfterAllKeys-3622a0bb-1cf4-4200-b830-5e937d57ac99";
```
Debug leftovers?

ayende replied:
No, they are being used, and I wanted to ensure that we are using values that no one else ever would

ayende committed:
RavenDB-20553 Now allow to define a compound field for indexes (not doing anything yet, just round tripping the information)

ayende committed:
RavenDB-20553 Generating automatic field from compound definition (early steps)

ayende committed:
RavenDB-20553 Small refactor to extract uncommon code outside of hot loop

ayende committed:
RavenDB-20553
- Adding support for most values in compound indexes
- For now, only supporting compound fields with 2 items

arekpalinski commented:
Can you please check failing test, do the rebase and solve the conflicts?

ayende committed:
RavenDB-20553
- Now compound indexes will use proper analyzer for the terms (but only a single term is supported)
- Can properly detect and optimize queries with simple equality and order by using a compound field

ayende commented:
Those failing tests are all about sorting nulls and Numerics.
The problem is that for null, we aren't ever _writing_ a value to the Numerics, so it isn't even found.

I can't think of an easy way to fix that which won't make things worse (have holes).
We can add a specification that a certain field is numeric, and determine if the value should be stored and how it should be there.

That said, we have the _exact_ same problem for strings, and we need to make a decision.

What I can do is to add a configuration for null indexing for fields.
But that leads to problems with scenarios such as:

`from Users order by Age` versus `from Users order by Age as long`

Both will use the same auto index, but we need to configure separately.

redknightlois commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/QueryOptimizer/CoraxBooleanItem.cs:
```diff
@@ -104,6 +108,92 @@ private CoraxBooleanItem(IndexSearcher searcher, Index index, FieldMetadata fiel
+                // }
+                //
+                // return baseMatch;
+        }
```
Leftovers or something that needs to be implemented yet?

ayende replied:
Meant to be implemented by @maciejaszyk , who took over this

ayende committed:
RavenDB-21137 Optimize compound sorting by fetching the first N+1 terms, then sorting just those, instead of the entire collection
- Allowing to limit the number of terms scanned during range operation

ayende committed:
RavenDB-21137 Fixing handling over very large terms in Corax when running equality comparisons

ayende committed:
RavenDB-21137  Fixing Studio compilation

ayende committed:
RavenDB-21137 Avoid allocating a string and use a char buffer instead

ayende committed:
RavenDB-21137 Avoid allocating a string reader per text

ayende committed:
RavenDB-21137 Actually use a span instead of char to avoid nastiness in allocations

ayende committed:
RavenDB-21131 When writing compound fields values, postfix with offsets

ayende committed:
RavenDB-21137 Moving from tree based to lookup persistent structure for managing container page lists, also managing the modified state _in memory_ and flushing on transaction commit only

ayende committed:
RavenDB-21137
- In memory "last container by page level metadata" to avoid searching too often
- Vectorized GetRequiredSizeAndPosition

ayende committed:
RavenDB-21137
- Remember the field root page instead of doing a dictionary lookup on each term

ayende committed:
RavenDB-21137 Avoid using a dictionary lookup (called millions of time) during recording of entries, instead, remember the index and use that to get the relevant entry

ayende committed:
RavenDB-21137 Caching Term & Offset attributes from the analyzer to avoid dictionary lookups

ayende committed:
RavenDB-21131 Adding tests & fixing comments for compound fields

ayende committed:
RavenDB-21131 Allowing to optimize compound filter & sort

ayende committed:
RavenDB-21131 Fixing compound field queries that used multiple sort fields

arekpalinski commented:
test this in debug please

ppekrol commented on src/Raven.Client/Documents/Queries/Timings/QueryTimings.cs:
```diff
@@ -10,6 +11,8 @@ public sealed class QueryTimings : IFillFromBlittableJson, IDynamicJson
 
         public IDictionary<string, QueryTimings> Timings { get; set; }
+        
+        public object QueryPlan { get; set; }
```
object?

ayende replied:
Changed to `IDynamicJson`, we really don't want to have a limit of what we can put here.

arekpalinski replied:
Done in https://github.com/ravendb/ravendb/pull/17176/commits/04d16c6516dd9974c952589f8d049b5907c90e0d

ppekrol commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -873,6 +873,13 @@ private static async Task<(long NumberOfResults, long TotalDocumentsSizeInBytes)
                 writer.WriteComma();
                 writer.WritePropertyName(nameof(result.Timings));
                 writer.WriteQueryTimings(context, result.Timings);
+                if (result.Timings.QueryPlan is IDynamicJson j)
```
I do not like that, can we have a dedicated type here?

ppekrol commented on test/FastTests/Corax/CompoundFieldsOnIndex.cs:
```diff
@@ -0,0 +1,155 @@
+                from u in users 
+                select new { u.Name, u.Location, u.Birthday };
+
+            CompoundFields.Add(new[] { nameof(User.Name), nameof(User.Birthday) });
```
beside that, I would consider adding method
```
protected void CompoundField(string fieldName1, string fieldName2)
```
with validation, etc

ppekrol replied:
also probably
```
protected void CompoundField(Expression<Func<TReduceResult, object>> field1, Expression<Func<TReduceResult, object>> field2)
```

ayende replied:
Good idea, done

arekpalinski replied:
Done in https://github.com/ravendb/ravendb/pull/17176/commits/04d16c6516dd9974c952589f8d049b5907c90e0d

ppekrol commented on test/FastTests/Corax/CompoundFieldsOnIndex.cs:
```diff
@@ -0,0 +1,155 @@
+                select new { u.Name, u.Location, u.Birthday };
+
+            CompoundFields.Add(new[] { nameof(User.Name), nameof(User.Birthday) });
+            CompoundFields.Add(new[] { nameof(User.Location), nameof(User.Name) });
```
regarding the validation, we should add this to `ValidateStaticIndexAsync` method? So this will throw before we put?

ayende replied:
I would rather not do that on the client side. Since this is runtime checked anyway, we will have the `CompoundField` method with two args, so that is already checked.
And a way to add more if needed. We'll absolutely like to have more fields allowed in the future.

arekpalinski replied:
Done in https://github.com/ravendb/ravendb/pull/17176/commits/04d16c6516dd9974c952589f8d049b5907c90e0d

arekpalinski commented:
Closing this PR as all commits are included in https://github.com/ravendb/ravendb/pull/17090 which just got merged. @ayende Please check Pawel's comments
